ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/layered/SparseTransactions.java

```diff
@@ -126,36 +125,86 @@ protected void internalReplaced(final PendingTransaction replacedTx) {
   @Override
   protected void internalBlockAdded(final BlockHeader blockHeader, final FeeMarket feeMarket) {}
 
+  /**
+   * We only want to promote transactions that have gap == 0, so there will be no gap in the prev
+   * layers. A promoted transaction is removed from this layer, and the gap data is updated for its
+   * sender.
+   *
+   * @param promotionFilter the prev layer's promotion filter
+   * @param freeSpace max amount of memory promoted txs can occupy
+   * @param freeSlots max number of promoted txs
+   * @return a list of transactions promoted to the prev layer
+   */
   @Override
-  public PendingTransaction promote(final Predicate<PendingTransaction> promotionFilter) {
-    final PendingTransaction promotedTx =
-        orderByGap.get(0).stream()
-            .map(txsBySender::get)
-            .map(NavigableMap::values)
-            .flatMap(Collection::stream)
-            .filter(promotionFilter)
-            .findFirst()
-            .orElse(null);
-
-    if (promotedTx != null) {
-      final Address sender = promotedTx.getSender();
-      final var senderTxs = txsBySender.get(sender);
-      senderTxs.pollFirstEntry();
-      processRemove(senderTxs, promotedTx.getTransaction(), PROMOTED);
-      if (senderTxs.isEmpty()) {
-        txsBySender.remove(sender);
-        orderByGap.get(0).remove(sender);
-        gapBySender.remove(sender);
-      } else {
-        final long firstNonce = senderTxs.firstKey();
-        final int newGap = (int) (firstNonce - (promotedTx.getNonce() + 1));
-        if (newGap != 0) {
-          updateGap(sender, 0, newGap);
+  public List<PendingTransaction> promote(
+      final Predicate<PendingTransaction> promotionFilter,
+      final long freeSpace,
+      final int freeSlots) {
+    long accSpace = 0;
```

Comment:
usedSpace?

Reply:
allocatedSpace?

Reply:
here `acc` stays for accumulated space, to keep how much space has been accumulated by the selected txs till now, do you think it is better to rename it?

Reply:
accumulated also works

Reply:
renaming to `accumulatedSpace`